File: Easy/1. Two Sum/1. Two Sum.py
from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:  # type: ignore
        ans = {}
        for i in range(len(nums)):
            el = nums[i]
            if target - el in ans:
                return [ans[target - el], i]
            ans[el] = i
        return []


""" <================= alt ans =================> """
"""
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:  # type: ignore
        ans = {}
        for i, el in enumerate(nums):
            remain = target - value

            if remain in ans:
                return [ans[remain], i]
            ans[el] = i
        return []
"""

File: Easy/1. Two Sum/1. Two Sum.rs
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        use std::collections::HashMap;

        let mut num_to_index = HashMap::new();

        for (index, num) in nums.iter().enumerate() {
            let complement = target - num;
            if let Some(complement_index) = num_to_index.get(&complement) {
                return vec![*complement_index as i32, index as i32];
            }
            num_to_index.insert(num, index);
        }

        panic!("No two elements add up to the target.");
    }
}
